hive/runners/base.py

```diff
@@ -108,17 +125,7 @@ def run_one_step(self, observation, turn, episode_metrics):
                 self._experiment_manager.update_step() or self._save_experiment
             )
 
-    def run_end_step(self, episode_metrics, done):
-        """Run the final step of an episode.
-
-        Args:
-            episode_metrics (Metrics): Keeps track of metrics for current episode.
-            done (bool): Whether this step was terminal.
-
-        """
-        return NotImplementedError
-
-    def run_episode(self):
+    def run_episode(self, environment):
```

Comment:
It took me a while to realize that the base runner only runs `run_testing` at the end of the training. And basically, if one wants to have an evaluation at the exact step that they like, one needs to (first know this logic and second) make a call in the `run_episode` function.

I predict that someone who writes a new runner function would miss this logic. Subsequently, they'll run their code only to realize that there is only one test at the end (😅). And finally, they manage to fix their issue by looking at either Single or Multi - agent runners.


Personally, I would not mind waiting until a training episode finishes and then I'll have my evaluation. Given the large number of training steps, an evaluation in step $k$ or $k + \epsilon, (k \ggg \epsilon)$ is perfectly fine and the final results/plots not only should be the same but also unrecognizable. 

On the other hand, I get that running testing at the exact step could be important. So, I am wondering if there is a better way to do this, maybe in the `update_runner_state`. Or otherwise, I think we should be super clear in RLHive's Document on this matter.

Reply:
I think updating the documentation is probably the best way to do this. Different runners will likely have different requirements about when to run the testing phase, and trying to enforce it in the base class would likely mess things up.

Reply:
Actually I think you are right, I looked at the code again, and the reason we weren't directly calling the save or testing in the update_state method is because it would cause issues with running testing in the middle of a training episode. Since this is no longer an issue, I'll move it to update_state in the stateless_agent pr.

Reply:
Take a look at #316 . Does that resolve this issue?